x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -315,10 +321,19 @@ private void onSearchResponse(SearchResponse searchResponse) {
             if (iterationResult.isDone()) {
                 logger.debug("Finished indexing for job [" + getJobId() + "], saving state and shutting down.");
 
-                // Change state first, then try to persist. This prevents in-progress
-                // STOPPING/ABORTING from
-                // being persisted as STARTED but then stop the job
-                doSaveState(finishAndSetState(), position.get(), this::onFinish);
+                // execute finishing tasks
+                try {
+                    onFinish(() -> {
+                        // Change state first, then try to persist. This prevents in-progress
+                        // STOPPING/ABORTING from
+                        // being persisted as STARTED but then stop the job
+                        doSaveState(finishAndSetState(), position.get(), () -> {});
+                    });
+                } catch (Exception e) {
```

Comment:
I think we should rather document the fact that the Runnable provided with `onFinish` (and maybe `onFailure`) should always run (even on errors). If we don't expect async calls in onFinish we can also keep the current code and just switch the execution order of `onFinish` and `doSaveState`.